applications/rag/frontend/container/rag_langchain/rag_chain.py

```diff
@@ -0,0 +1,132 @@
+        table_name = CHAT_HISTORY_TABLE_NAME
+    )
+    # history = None
+    # if session_id in chat_history_map:
+    #     history = chat_history_map[session_id]
+    # else:
+    #     history = PostgresChatMessageHistory.create_sync(
+    #         engine,
+    #         session_id=session_id,
+    #         table_name = CHAT_HISTORY_TABLE_NAME
+    #     )
+    #     chat_history_map[session_id] = history
+    print(f"Retrieving history for session {session_id} with {len(history.messages)} items:")
+    print("\t" + "\n\t".join([message.content for message in history.messages]))
+    return history
+
+def clear_chat_history(session_id: str):
+    history = PostgresChatMessageHistory.create_sync(
+        engine,
+        session_id=session_id,
+        table_name = CHAT_HISTORY_TABLE_NAME
+    )    # if session_id in chat_history_map:
+    #     history = chat_history_map[session_id]
+    history.clear()
+        # del chat_history_map[session_id]
+
+#TODO: limit number of tokens in prompt to MAX_INPUT_LENGTH 
+# (as specified in hugging face TGI input parameter)
+
+def create_chain() -> RunnableWithMessageHistory: 
+    # TODO HuggingFaceTextGenInference class is deprecated. 
+    # The warning is: 
+    #                The class `langchain_community.llms.huggingface_text_gen_inference.HuggingFaceTextGenInference` 
+    #                was deprecated in langchain-community 0.0.21 and will be removed in 0.2.0. Use HuggingFaceEndpoint instead
+    # The replacement is HuggingFace Endoint, which requires a huggingface
+    # hub API token. Either need to add the token to the environment, or need to find a method to call TGI
+    # without the token.
+    # Example usage of HuggingFaceEndpoint:
+    # llm = HuggingFaceEndpoint(
+    #                 endpoint_url=f'http://{INFERENCE_ENDPOINT}/',
+    #                 max_new_tokens=512,
+    #                 top_k=10,
+    #                 top_p=0.95,
+    #                 typical_p=0.95,
+    #                 temperature=0.01,
+    #                 repetition_penalty=1.03,
+    #                 huggingfacehub_api_token="my-api-key"
+    #             )
+    model = HuggingFaceTextGenInference(
+        inference_server_url=f'http://{INFERENCE_ENDPOINT}/',
+        max_new_tokens=512,
+        top_k=10,
+        top_p=0.95,
+        typical_p=0.95,
+        temperature=0.01,
+        repetition_penalty=1.03,
+    )
+
+    langchain_embed = HuggingFaceEmbeddings(model_name=SENTENCE_TRANSFORMER_MODEL)
+    vector_store = CustomVectorStore(langchain_embed, init_connection_pool(Connector()))
+    retriever = vector_store.as_retriever()
+
+    setup_and_retrieval = RunnableParallel(
+        {
+            "context": retriever, 
+            QUESTION: RunnableLambda(lambda d: d[QUESTION]), 
+            HISTORY: RunnableLambda(lambda d: d[HISTORY])
+        }
+    )
+    chain = setup_and_retrieval | prompt | model
+    chain_with_history = RunnableWithMessageHistory(
```

Comment:
Maybe https://python.langchain.com/docs/langgraph/ is a better choice

Reply:
I hadn't seen it before. Where do you think we could use a stateful actor model? I don't think I'll have time to look into it before this PR goes out.

Reply:
We can do this later, I think as infra we should support more flexibility.
for example, user may want to retrieve data from prod db and experiment db. or like different team will maintain different domian knowledge db.
And also for chat history we can have the conditional edge, if chat history existing, we can do something like re-rank the result from the cosin similarity search. 
I agree this PR should focus on the chat history.